### ISSUE TITLE
psutil.containertype or something like that ?

### ISSUE THREAD

**Issue by bitranox** [giampaolo/psutil#1511]
psutil should report in which (if any) container we are running - 
like Heroku, LXC, Docker, VMWare, etc.

probably additional details like Version and other useful information can be added

**Comment by giampaolo** [comment]
I think this may be a good idea. I have close to zero experience re. containers meaning this needs to be investigated. :)
In terms of API I think it makes sense to return a string similarly to ‘os.name’ and ‘sys.platform’:

‘’’python
>>> psutil.container()
“docker”
‘’’

Including the version probably means looking for troubles as it would be too dependent from the passing of third-parties CLI tools output.

If you want to investigate this topic fee free to use this ticket as a

**Comment by bitranox** [comment]
I think we can start with something like : 

psutil.container.is_container() -> bool

this should be easy on linux - because some (mount?) operations are simply not possible inside containers - no idea for windows ;-)

then You can gradually add container types: 

psitil.container.type() -> str

and throw an Exception like PsutilContainerDetectionError if You can not determine the Type.

This makes it possible to check if we are inside a container easy (inside Linux) and if someone really needs the type - well it might work (or not) - since container landscapes are evolving rapidly nowadays.

Are You back from China ? I hope You had a pleasant stay ... 

yours sincerely

Robert

**Comment by giampaolo** [comment]
Nope, still in China, and I will be for the next 1-2 months probably. =)

> I think we can start with something like :
> psutil.container.is_container() -> bool

That would mean adding a new namespace (a first in psutil). I think it's safer to aim at the minimal possible API. In my head that means a  function returning either a string or `None` in case you're not in a container or the container cannot be determined. That would basically be the same as `psutil.cpu_count()` (which mimicks `os.cpu_count()`). As such you'll just do:

```python
if psutil.container() == 'docker':
    ...
```

...as opposed to:

```python
try:
    is_docker = psutil.container.type()
except psutil.container.SomeNewException:
    is_docker = False
if is_docker:
    ...
```

I think the main headache here is determining if we're on docker/vmware/etc. on different platforms (other than Linux).

**Comment by bitranox** [comment]
hm - I disagree here.

It is very easy to determine IF we are in a container, so we should have two functions : 

```
psutil.container.inside_container() -> bool  #  very reliable to detect, gives True or False

psutil.container.type() -> [str]   #  not very reliable to detect, should raise Exception if detection not possible
```

In general, Exceptions are preferable against returning 0 or None - 
because returning such a value forces You to store and/or test the return value immideately.

I think Your example is not fair here, there can be many situautions if You need to test additional 
if the return value is None.

And almost every production code is wrapped in try - except at some level - so You will catch the Exception there and then decide what to do with it.

If You want to stick to Your idea, I would propose a return value "unknown" # type: str
what should mean : Yes, we run inside a container of undetected type.

I would still prefer an Exception if there is an error - thats what Exceptions are for ;-)

in my opinion it is just much more obviouse if I receive an 
psutil.container.ContainerTypeDetectionError then a return value of "None" which can mean that we are not inside a container, or can mean that we are not able to detect the container type .... 

> I think the main headache here is determining if we're on docker/vmware/etc. on different platforms (other than Linux).

yes - exactly - another argument for 
psutil.container.ContainerTypeDetectionError
psutil.container.ContainerDetectionError

you can start to integrate what is easy - and throw exceptions in other cases without the misleading "None" return value !

have a good time, Xie Xie !

Robert
